cmd/registry-experimental/cmd/generate/openapi.go

```diff
@@ -0,0 +1,226 @@
+	} else if core.IsProto(spec.GetMimeType()) && core.IsZipArchive(spec.GetMimeType()) {
+		log.FromContext(ctx).Debugf("Computing %s/specs/%s", spec.Name, relation)
+		openapi, err = openAPIFromZippedProtos(ctx, spec.Name, data)
+		if err != nil {
+			return fmt.Errorf("error processing protos %s: %s", spec.Name, err)
+		}
+	} else if core.IsDiscovery(spec.GetMimeType()) {
+		log.FromContext(ctx).Debugf("Computing %s/specs/%s", spec.Name, relation)
+		openapi, err = openAPIFromDiscovery(spec.Name, data)
+		if err != nil {
+			return fmt.Errorf("error processing discovery %s: %s", spec.Name, err)
+		}
+	} else {
+		return fmt.Errorf("we don't know how to generate OpenAPI for %s", spec.Name)
+	}
+	specName, _ := names.ParseSpec(spec.GetName())
+	messageData := []byte(openapi)
+	messageData, err = core.GZippedBytes(messageData)
+	if err != nil {
+		return fmt.Errorf("failed to compress generated OpenAPI: %s", err)
+	}
+	newSpec := &rpc.ApiSpec{
+		Name:     specName.Version().Spec(relation).String(),
+		MimeType: "application/x.openapi+gzip;version=3",
+		Contents: messageData,
+		Filename: "openapi.yaml",
+	}
+	_, err = task.client.UpdateApiSpec(ctx, &rpc.UpdateApiSpecRequest{
+		ApiSpec:      newSpec,
+		AllowMissing: true,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to generate OpenAPI: %s", err)
+	}
+	return nil
+}
+
+// openAPIFromZippedProtos runs the OpenAPI generator and returns the results.
+// This uses protoc and https://github.com/google/gnostic/tree/master/cmd/protoc-gen-openapi
+// which can be installed using
+//   go install github.com/google/gnostic/cmd/protoc-gen-openapi@latest
+func openAPIFromZippedProtos(ctx context.Context, name string, b []byte) (string, error) {
+	// create a tmp directory
+	root, err := ioutil.TempDir("", "registry-protos-")
+	if err != nil {
+		return "", err
+	}
+	// whenever we finish, delete the tmp directory
+	defer os.RemoveAll(root)
+	// unzip the protos to the temp directory
+	_, err = core.UnzipArchiveToPath(b, root+"/protos")
+	if err != nil {
+		return "", err
+	}
+	return generateOpenAPIForDirectory(ctx, name, root)
+}
+
+func generateOpenAPIForDirectory(ctx context.Context, name string, root string) (string, error) {
+	// run protoc on all of the protos in the main directory
+	protos := []string{}
+	err := filepath.Walk(root+"/protos",
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if strings.HasSuffix(path, ".proto") {
+				protos = append(protos, strings.TrimPrefix(path, root+"/"))
+			}
+			return nil
+		})
+	if err != nil {
+		return "", err
+	}
+	parts := []string{}
+	parts = append(parts, protos...)
+	parts = append(parts, "-I")
+	parts = append(parts, "protos")
+	parts = append(parts, "--openapi_out=.")
+	cmd := exec.Command("protoc", parts...)
+	cmd.Dir = root
+	log.FromContext(ctx).Debugf("Running %+v\n", cmd)
+	data, err := cmd.CombinedOutput()
+	if err != nil {
+		log.FromContext(ctx).Errorf("error %+v\n", err)
```

Comment:
Should we emit the instructions to install that are above here?

Reply:
Done (and below). The logging is messy but the links are in there.

```
 % registry-experimental generate openapi projects/omnibus/locations/global/apis/googleapis.com-apigeeregistry/versions/v1/specs/-
  INFO[0000] generate openapi for projects/omnibus/locations/global/apis/googleapis.com-apigeeregistry/versions/v1/specs/discovery uid=efade7b0
 DEBUG[0001] Computing projects/omnibus/locations/global/apis/googleapis.com-apigeeregistry/versions/v1/specs/discovery/specs/generated uid=efade7b0
 ERROR[0001] See "https://www.npmjs.com/package/api-spec-converter" for installation instructions error=exec: "api-spec-converter": executable file not found in $PATH uid=efade7b0
  WARN[0001] Task failed: generate openapi for projects/omnibus/locations/global/apis/googleapis.com-apigeeregistry/versions/v1/specs/discovery error=error processing discovery projects/omnibus/locations/global/apis/googleapis.com-apigeeregistry/versions/v1/specs/discovery: exec: "api-spec-converter": executable file not found in $PATH uid=efade7b0
  INFO[0001] generate openapi for projects/omnibus/locations/global/apis/googleapis.com-apigeeregistry/versions/v1/specs/google-cloud-apigeeregistry-v1 uid=efade7b0
 DEBUG[0001] Computing projects/omnibus/locations/global/apis/googleapis.com-apigeeregistry/versions/v1/specs/google-cloud-apigeeregistry-v1/specs/generated uid=efade7b0
 DEBUG[0001] Running /usr/local/bin/protoc protos/google/api/annotations.proto protos/google/api/client.proto protos/google/api/field_behavior.proto protos/google/api/http.proto protos/google/api/httpbody.proto protos/google/api/launch_stage.proto protos/google/api/resource.proto protos/google/cloud/apigeeregistry/v1/provisioning_service.proto protos/google/cloud/apigeeregistry/v1/registry_models.proto protos/google/cloud/apigeeregistry/v1/registry_service.proto protos/google/longrunning/operations.proto protos/google/rpc/status.proto -I protos --openapi_out=.
 uid=efade7b0
 ERROR[0001] See "https://github.com/google/gnostic/tree/main/cmd/protoc-gen-openapi" for installation instructions error=exit status 1 uid=efade7b0
  WARN[0001] Task failed: generate openapi for projects/omnibus/locations/global/apis/googleapis.com-apigeeregistry/versions/v1/specs/google-cloud-apigeeregistry-v1 error=error processing protos projects/omnibus/locations/global/apis/googleapis.com-apigeeregistry/versions/v1/specs/google-cloud-apigeeregistry-v1: exit status 1 uid=efade7b0
  INFO[0001] generate openapi for projects/omnibus/locations/global/apis/googleapis.com-apigeeregistry/versions/v1/specs/openapi uid=efade7b0
```